docs/extending/localtunneldebugger.md

```diff
@@ -0,0 +1,122 @@
+
+### Implementing the debugger
+
+Your provider is responsible for resolving the debug target (such as a pod or service)
+when a resource is targeted via the Cluster explorer. You can use the cluster explorer
+API to do so (resolveCommandTarget method). This parameter will be undefined if the debugging
+action is initiated through the Visual Studio command palette. Your provider controls
+the tunneling and debugging behavior of the command.
+
+## Registering the debug provider
+
+In order to hook into the Debug (Local Tunnel) option, a local debug provider must be _registered_
+with the Kubernetes extension. This is the responsibility of the VS Code extension that hosts
+the debug provider. To do this, the extension must:
+
+* Activate in response to the `kubernetes.cloudExplorer` view
+* Activate in response to the `Debug (Local Tunnel)` command
+* Request the Kubernetes extension's Local Tunnel Debug Provider API
+* Call the `register` method for each cloud provider it wants to display
+
+### Activating the debug provider extension
+
+Your extension needs to activate in response to the cluster explorer and command palette
+commands, so that it can register as a debug provider before the debug session is started.
+To do this, your `package.json` must include the following activation events:
+
+```json
+    "activationEvents": [
+        "onView:extension.vsKubernetesExplorer",
+        "onCommand:extension.vsKubernetesDebugLocalTunnel",
+    ],
+```
+
+Depending on your extension you may have other activation events as well.
+
+### Registering local tunnel debug providers with the Kubernetes extension
+
+In your extension's `activate` function, you must register your provider(s) using the
+Kubernetes extension API.  The following sample shows how to do this using the NPM
+helper package; if you don't use the helper then the process of requesting the API is
+more manual but the registration is the same.
+
+```javascript
+const MY_PROVIDER = {
+    id: "Contoso Trampoline Debugger",
+    startDebugging: (target?: any) => startDebugSession(target)
+};
+
+export async function activate(context: vscode.ExtensionContext) {
+    const localTunnelDebugger = await k8s.extension.localTunnelDebugger.v1;
+
+    if (!localTunnelDebugger.available) {
+        console.log("Unable to register provider: " + localTunnelDebugger.reason);
+        return;
+    }
+
+    localTunnelDebugger.api.registerLocalTunnelDebugProvider(MY_PROVIDER);
+}
+```
+
+Your debug provider is now ready for testing!
+
+## Tagging your debug provider extension
+
+The Local Tunnel Debugger has a link for users to find providers on the Visual Studio
+Marketplace.  If you'd like to appear in this link, please tag your extension
+with the keyword `kubernetes-extension-local-tunnel-provider` in `package.json`.
+For example:
+
+```json
+{
+    "keywords": [
+        "kubernetes",
+        "contoso",
+        "megafloof",
+        "kubernetes-extension-local-tunnel-provider"
```

Comment:
Should the tag be `kubernetes-extension-local-tunnel-debug-provider`? This seems to be the name used in the rest of the file.

Reply:
...or `kubernetes-extension-local-tunnel-debugger`? Seems like we're using this name as well.
Are there any conceptual differences between both?

Reply:
Yeah, the **debug provider** would be the extension that actually provides the debugging functionality, while the **debugger** is the layer that keeps track of the debug providers and connects the Debug (Local Tunnel) command to the provider's startDebugging API. (I tried to copy the way the CloudExplorer/CloudProvider interfaces interact.) In line with this, I'll go with "kubernetes-extension-local-tunnel-debug-provider" :)